File: tests/adapters/pi/ignore.test.ts
import { describe, expect, it } from 'vitest';

import { globMatch, matchesIgnore } from '../../../src/adapters/pi/ignore.js';
import { DEFAULT_PI_ADAPTER } from '../../../src/adapters/pi/defaults.js';

describe('globMatch — 冻结的 glob 语义', () => {
  // 字面量 / '*' 非跨 '/' / 尾 '/' 目录前缀 —— 逐条锁死
  const cases: { pattern: string; path: string; expected: boolean; note: string }[] = [
    { pattern: 'auth.json', path: 'auth.json', expected: true, note: '字面量精确匹配' },
    { pattern: 'auth.json', path: 'sub/auth.json', expected: false, note: '字面量不匹配子路径' },
    { pattern: 'auth.json', path: 'auth.jsonx', expected: false, note: '字面量必须全等' },
    { pattern: '*.bak', path: 'models.json.bak', expected: true, note: '* 任意非 / 字符' },
    { pattern: '*.bak', path: 'a/b.bak', expected: false, note: '* 不跨 /' },
    { pattern: '*.bak-*', path: 'models.json.bak-predirect', expected: true, note: '两侧 *' },
    { pattern: '*.bak-*', path: 'x/y.bak-predirect', expected: false, note: '两段均不跨 /' },
    { pattern: '*.bak-*', path: 'models.json.bak', expected: false, note: '缺 -<后缀>' },
    { pattern: '*.log', path: 'pi-tui-crash.log', expected: true, note: '日志后缀' },
    { pattern: '*cache*', path: 'pi-foo-cache', expected: true, note: '两端通配' },
    { pattern: '*cache*', path: 'mycache.json', expected: true, note: '中缀匹配' },
    { pattern: '*cache*', path: 'keep/index.js', expected: false, note: '不命中' },
    { pattern: 'sessions/', path: 'sessions/x', expected: true, note: '尾 / = 目录前缀' },
    { pattern: 'sessions/', path: 'sessions/a/b/c', expected: true, note: '前缀命中深层' },
    { pattern: 'sessions/', path: 'sessions', expected: true, note: '目录本身命中' },
    { pattern: 'sessions/', path: 'mysessions/x', expected: false, note: '前缀必须整段' },
    { pattern: 'npm/', path: 'npm', expected: true, note: '目录自身' },
    { pattern: 'npm/', path: 'npm/foo/bar', expected: true, note: '深层' },
    { pattern: 's*', path: 'skills/foo', expected: false, note: '单通配不跨 /' },
    { pattern: 's*', path: 'settings.json', expected: true, note: '单段通配' },
    { pattern: 's*/foo', path: 'skills/foo', expected: true, note: '多段通配' },
    { pattern: '', path: 'anything', expected: false, note: '空模式不匹配' },
  ];

  it.each(cases)('$pattern vs $path → $expected ($note)', ({ pattern, path, expected }) => {
    expect(globMatch(path, pattern)).toBe(expected);
  });

  it('不支持 ? 与 **（按字面/单星处理，不跨 /）', () => {
    // '?' 是字面字符
    expect(globMatch('a?b', 'a?b')).toBe(true);
    expect(globMatch('axb', 'a?b')).toBe(false);
    // '**' 等价单个 '*'，仍不跨 '/'（只吃一段）
    expect(globMatch('a/b', '**')).toBe(false);
    expect(globMatch('ab', '**')).toBe(true);
    expect(globMatch('deep/x', '**/x')).toBe(true);
    expect(globMatch('a/b/x', '**/x')).toBe(false);
  });

  it('前导 ./ 与 / 被归一化', () => {
    expect(globMatch('./sessions/x', 'sessions/')).toBe(true);
    expect(globMatch('/sessions/x', 'sessions/')).toBe(true);
    expect(globMatch('sessions/x', './sessions/')).toBe(true);
  });
});

describe('matchesIgnore — 表驱动（默认 pi ignore 列表）', () => {
  const ignore = DEFAULT_PI_ADAPTER.ignore ?? [];

  const cases: { path: string; expected: boolean; note: string }[] = [
    { path: 'auth.json', expected: true, note: '显式列名' },
    { path: 'trust.json', expected: true, note: '显式列名' },
    { path: 'sessions/x', expected: true, note: 'sessions/ 前缀' },
    { path: 'sessions/2026/foo.jsonl', expected: true, note: 'sessions/ 深层' },
    { path: 'npm/y', expected: true, note: 'npm/ 前缀' },
    { path: 'git/config', expected: true, note: 'git/ 前缀' },
    { path: 'tmp/scratch', expected: true, note: 'tmp/ 前缀' },
    { path: 'bin/homer', expected: true, note: 'bin/ 前缀' },
    { path: 'models.json.bak-predirect', expected: true, note: '*.bak-*' },
    { path: 'models.json.bak', expected: true, note: '*.bak' },
    { path: 'settings.json.bak', expected: true, note: '*.bak' },
    { path: 'pi-tui-crash.log', expected: true, note: '*.log' },
    { path: 'run-history.jsonl', expected: true, note: '显式列名' },
    { path: 'settings.json', expected: false, note: '正常配置' },
    { path: 'skills/foo/SKILL.md', expected: false, note: '正常配置' },
    { path: 'themes/dark.json', expected: false, note: '正常配置' },
    { path: 'sessions.json', expected: false, note: '前缀模式不误伤同前缀文件' },
    { path: 'auth.json.example', expected: false, note: '字面量不误伤' },
    { path: 'models/bak', expected: false, note: '*.bak 不跨 /' },
    { path: 'nested/pi-tui-crash.log', expected: false, note: '*.log 不跨 /（仅顶层段）' },
  ];

  it.each(cases)('$path → $expected ($note)', ({ path, expected }) => {
    expect(matchesIgnore(path, ignore)).toBe(expected);
  });

  it('空 pattern 列表 → 恒 false', () => {
    expect(matchesIgnore('anything', [])).toBe(false);
  });

  it('自定义模式列表同样生效', () => {
    expect(matchesIgnore('foo/bar.txt', ['foo/'])).toBe(true);
    expect(matchesIgnore('foo/bar.txt', ['*.txt'])).toBe(false);
    expect(matchesIgnore('bar.txt', ['*.txt'])).toBe(true);
  });
});

File: internal/adapter/ignore_test.go
package adapter_test

import (
	"testing"

	"github.com/zzjcool/homer-cli/internal/adapter"
)

func TestGlobMatchFrozenTable(t *testing.T) {
	tests := []struct {
		name     string
		pattern  string
		path     string
		expected bool
	}{
		{name: "literal exact", pattern: "auth.json", path: "auth.json", expected: true},
		{name: "literal not nested", pattern: "auth.json", path: "sub/auth.json", expected: false},
		{name: "literal full", pattern: "auth.json", path: "auth.jsonx", expected: false},
		{name: "star suffix", pattern: "*.bak", path: "models.json.bak", expected: true},
		{name: "star does not cross slash", pattern: "*.bak", path: "a/b.bak", expected: false},
		{name: "two stars", pattern: "*.bak-*", path: "models.json.bak-predirect", expected: true},
		{name: "missing suffix", pattern: "*.bak-*", path: "models.json.bak", expected: false},
		{name: "star both sides", pattern: "*cache*", path: "mycache.json", expected: true},
		{name: "nested cache not a single path", pattern: "*cache*", path: "keep/index.js", expected: false},
		{name: "directory prefix child", pattern: "sessions/", path: "sessions/x", expected: true},
		{name: "directory prefix deep child", pattern: "sessions/", path: "sessions/a/b/c", expected: true},
		{name: "directory itself", pattern: "sessions/", path: "sessions", expected: true},
		{name: "directory boundary", pattern: "sessions/", path: "mysessions/x", expected: false},
		{name: "prefix star", pattern: "s*", path: "settings.json", expected: true},
		{name: "prefix star does not cross slash", pattern: "s*", path: "skills/foo", expected: false},
		{name: "multi-segment star", pattern: "s*/foo", path: "skills/foo", expected: true},
		{name: "question is literal", pattern: "a?b", path: "a?b", expected: true},
		{name: "question is not wildcard", pattern: "a?b", path: "axb", expected: false},
		{name: "double star is still one segment", pattern: "**", path: "ab", expected: true},
		{name: "double star does not cross slash", pattern: "**", path: "a/b", expected: false},
		{name: "double star nested one level", pattern: "**/x", path: "deep/x", expected: true},
		{name: "double star nested two levels", pattern: "**/x", path: "a/b/x", expected: false},
		{name: "empty pattern", pattern: "", path: "anything", expected: false},
		{name: "leading dot slash", pattern: "sessions/", path: "./sessions/x", expected: true},
		{name: "leading slash", pattern: "sessions/", path: "/sessions/x", expected: true},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			if got := adapter.GlobMatch(test.path, test.pattern); got != test.expected {
				t.Fatalf("GlobMatch(%q, %q) = %v, want %v", test.path, test.pattern, got, test.expected)
			}
		})
	}
}

func TestMatchesIgnoreAndBareGlobGuard(t *testing.T) {
	patterns := []string{"auth.json", "sessions/", "*.bak-*", "*.log"}
	cases := map[string]bool{
		"auth.json":                 true,
		"sessions/2026/foo.jsonl":   true,
		"models.json.bak-predirect": true,
		"pi-tui-crash.log":          true,
		"nested/pi-tui-crash.log":   false,
		"sessions.json":             false,
		"auth.json.example":         false,
		"settings.json":             false,
		"skills/foo/SKILL.md":       false,
		"models.json.bak":           false,
	}
	for path, want := range cases {
		if got := adapter.MatchesIgnore(path, patterns); got != want {
			t.Errorf("MatchesIgnore(%q) = %v, want %v", path, got, want)
		}
	}
	if adapter.MatchesIgnore("anything", nil) {
		t.Fatal("nil ignore patterns must not match")
	}

	for _, pattern := range []string{"*", "**", "*/", "**/", "./*", "/*", ".//*", "/**/"} {
		if !adapter.IsBareGlob(pattern) {
			t.Errorf("IsBareGlob(%q) = false, want true", pattern)
		}
	}
	for _, pattern := range []string{"skills/*", "skills/*/inner", "skills/agent-browser", "a*/", ""} {
		if adapter.IsBareGlob(pattern) {
			t.Errorf("IsBareGlob(%q) = true, want false", pattern)
		}
	}
}
